File: java/ofdpdfsigner_jni/src/main/java/com/fzz/seal/OFDStamp.java
package com.fzz.seal;

import com.fzz.seal.callback.ISignCallBack;
import com.fzz.seal.callback.IVerifyCallBack;
import com.fzz.seal.jni.IJNI;
import com.fzz.seal.jni.ReturnData;

import java.nio.charset.StandardCharsets;

public class OFDStamp {

    public static final int RIDING_LEFT = 0;
    public static final int RIDING_RIGHT = 1;
    public static final int RIDING_TOP = 2;
    public static final int RIDING_BOTTOM = 3;
    public static final int RIDING_CUTTING_RIGHT = 4;
    public static final int RIDING_CUTTING_TOP = 5;
    /*
     OFD 位置盖章
     stampData：印章数据
     ofdData：OFD数据
     pageIndex：页码 0开始
     x：开始位置
     y：开始位置
     stampwidth：印章宽
     stampheight: 印章高
     isContinueSign：是否继续盖章
     signCallBack：签名回调
     返回：电子印章数据
     */
//    public static byte[] signStamp_Pos(byte[] stampData,byte[] ofdData, int pageIndex,double x,double y,
//                                          double stampwidth, double stampheight,boolean isContinueSign,
//                                          ISignCallBack signCallBack)
//    {
//        byte[] ret = null;
//        if ( stampData == null || stampData.length <= 0 || ofdData == null || ofdData.length <= 0 ||
//                stampwidth < 0 || stampheight < 0 || pageIndex < 0 ) {
//            return ret;
//        }
//        ReturnData returnData = new ReturnData();
//
//        IJNI.OFDSignStamp_Pos(stampData, ofdData,  pageIndex, x, y,stampwidth,  stampheight, isContinueSign,signCallBack,returnData);
//
//        if ( returnData.getErrorCode() == 0 && returnData.getResultData() != null ) {
//            byte[] temp = returnData.getResultData();
//            //System.out.println("zip len : "+temp.length);
//            ret = new byte[temp.length];
//            System.arraycopy(temp,0,ret,0,temp.length);
//        }
//        return ret;
//    }
    /*
     OFD 位置盖章
     stampData：印章数据
     ofdData：OFD数据
     pageIndex：页码 0开始
     x：开始位置
     y：开始位置
     stampwidth：印章宽
     stampheight: 印章高
     isContinueSign：是否继续盖章
     signCallBack：签名回调
     返回：电子印章数据
     */
    public static byte[] positionSign(long fileHandle,byte[] sealData, int pageNumber,double stampX,double stampY,
                                       double width, double height,boolean isLocked,
                                       ISignCallBack signCallBack)
    {
        byte[] ret = null;
        if ( sealData == null || sealData.length <= 0 || fileHandle == 0 ||
                width < 0 || width < 0 || pageNumber <= -2 || pageNumber == 0) {
            return ret;
        }
        ReturnData returnData = new ReturnData();

        IJNI.OFDSignStamp_Pos_Handle(fileHandle,sealData,  pageNumber, stampX, stampY,width,  height, !isLocked,signCallBack,returnData);

        if ( returnData.getErrorCode() == 0 && returnData.getResultData() != null ) {
            byte[] temp = returnData.getResultData();
            //System.out.println("zip len : "+temp.length);
            ret = new byte[temp.length];
            System.arraycopy(temp,0,ret,0,temp.length);
        }
        return ret;
    }
    /*
     OFD文件骑缝章签章
     stampData：印章数据
     ofdData：OFD数据
     RidingType：骑缝章类型
     stampwidth：印章宽
     stampheight: 印章高
     isContinueSign：是否继续盖章
     signCallBack：签名回调
     返回：电子印章数据
     */
//    public static byte[] signStamp_Riding(byte[] stampData,byte[] ofdData, int RidingType,
//                                          double stampwidth, double stampheight,boolean isContinueSign,ISignCallBack signCallBack)
//    {
//        byte[] ret = null;
//        if ( stampData == null || stampData.length <= 0 || ofdData == null || ofdData.length <= 0 ||
//                stampwidth < 0 || stampheight < 0  ) {
//            return ret;
//        }
//        ReturnData returnData = new ReturnData();
//
//        IJNI.OFDSignStamp_Riding(stampData, ofdData, RidingType,stampwidth,  stampheight,isContinueSign,signCallBack,returnData);
//
//        if ( returnData.getErrorCode() == 0 && returnData.getResultData() != null ) {
//            byte[] temp = returnData.getResultData();
//            //System.out.println("zip len : "+temp.length);
//            ret = new byte[temp.length];
//            System.arraycopy(temp,0,ret,0,temp.length);
//        }
//        return ret;
//    }
    /*
     OFD文件骑缝章签章
     stampData：印章数据
     ofdData：OFD数据
     RidingType：骑缝章类型
     stampwidth：印章宽
     stampheight: 印章高
     isContinueSign：是否继续盖章
     signCallBack：签名回调
     返回：电子印章数据
     */
    public static byte[] ridingSign(long fileHandle,byte[] sealData, String ridingType,
                                          double width, double height,boolean isLocked,ISignCallBack signCallBack)
    {
        byte[] ret = null;
        if ( sealData == null || sealData.length <= 0 || fileHandle == 0 ||
                width < 0 || height < 0  ) {
            return ret;
        }
        ReturnData returnData = new ReturnData();

        int iType = Integer.parseInt(ridingType);

        IJNI.OFDSignStamp_Riding_Handle(fileHandle,sealData, iType,width,  height,!isLocked,signCallBack,returnData);

        if ( returnData.getErrorCode() == 0 && returnData.getResultData() != null ) {
            byte[] temp = returnData.getResultData();
            //System.out.println("zip len : "+temp.length);
            ret = new byte[temp.length];
            System.arraycopy(temp,0,ret,0,temp.length);
        }
        return ret;
    }
    /*
     OFD文件关键字签章
     imageData：印章数据
     ofdData：OFD数据
     keyword：关键字
     KeywordRule：关键字规则
     stampwidth：印章宽
     stampheight: 印章高
     isContinueSign：是否继续盖章
     signCallBack：签名回调
     返回：电子印章数据
     */
//    public static byte[] signStamp_KeyWord(byte[] stampData, byte[] ofdData, String keyword,String KeywordRule,
//                                           double stampwidth, double stampheight,boolean isContinueSign,ISignCallBack signCallBack)
//    {
//        byte[] ret = null;
//        if ( stampData == null || stampData.length <= 0 || ofdData == null || ofdData.length <= 0 ||
//                stampwidth < 0 || stampheight < 0 || keyword == null || keyword.length() <= 0
//                || KeywordRule == null || KeywordRule.length() <= 0 ) {
//            return ret;
//        }
//        ReturnData returnData = new ReturnData();
//
//        IJNI.OFDSignStamp_KeyWord(stampData, ofdData,  keyword.getBytes(StandardCharsets.UTF_8), KeywordRule.getBytes(StandardCharsets.UTF_8), stampwidth,  stampheight,isContinueSign,signCallBack,returnData);
//
//        if ( returnData.getErrorCode() == 0 && returnData.getResultData() != null ) {
//            byte[] temp = returnData.getResultData();
//            //System.out.println("zip len : "+temp.length);
//            ret = new byte[temp.length];
//            System.arraycopy(temp,0,ret,0,temp.length);
//        }
//        return ret;
//    }

    /*
     OFD文件关键字签章
     imageData：印章数据
     ofdData：OFD数据
     keyword：关键字
     KeywordRule：关键字规则
     stampwidth：印章宽
     stampheight: 印章高
     isContinueSign：是否继续盖章
     signCallBack：签名回调
     返回：电子印章数据
     */
    public static byte[] keywordSign(long fileHandle,byte[] sealData, String keyword,String KeywordRule,
                                           double width, double height,boolean isLocked,ISignCallBack signCallBack)
    {
        byte[] ret = null;
        if ( sealData == null || sealData.length <= 0 || fileHandle == 0 ||
                width < 0 || height < 0 || keyword == null || keyword.length() <= 0
                || KeywordRule == null || KeywordRule.length() <= 0 ) {
            return ret;
        }
        ReturnData returnData = new ReturnData();

        IJNI.OFDSignStamp_KeyWord_Handle(fileHandle,sealData,  keyword.getBytes(StandardCharsets.UTF_8), KeywordRule.getBytes(StandardCharsets.UTF_8), width,  height,!isLocked,signCallBack,returnData);

        if ( returnData.getErrorCode() == 0 && returnData.getResultData() != null ) {
            byte[] temp = returnData.getResultData();
            //System.out.println("zip len : "+temp.length);
            ret = new byte[temp.length];
            System.arraycopy(temp,0,ret,0,temp.length);
        }
        return ret;
    }
    /*
     OFD 文档验证
     docdata：OFD数据
     IVerifyCallBack：验证回调
     返回：验证结果，json数据
     */
    public static String verify(byte[] fileData, IVerifyCallBack verifyCallBack)
    {
        String ret = "";
        if ( fileData == null || fileData.length <= 0 ) {
            return ret;
        }
        ReturnData returnData = new ReturnData();

        IJNI.OFDVerifyDocment(fileData,verifyCallBack,returnData);

        if ( returnData.getErrorCode() == 0 && returnData.getResultData() != null ) {
            byte[] temp = returnData.getResultData();
            ret = new String(temp);
        }
        return ret;
    }

    /*
     OFD 删除文档中全部签名数据
     fileHandle：OFD 文档句柄
     返回：删除后文档数据
     */
    public static byte[] clearSigns(long fileHandle)
    {
        byte[] ret = null;
        if ( fileHandle == 0 ) {
            return ret;
        }
        ReturnData returnData = new ReturnData();

        IJNI.OFDClearStamps(fileHandle,returnData);

        if ( returnData.getErrorCode() == 0 && returnData.getResultData() != null ) {
            byte[] temp = returnData.getResultData();
            //System.out.println("zip len : "+temp.length);
            ret = new byte[temp.length];
            System.arraycopy(temp,0,ret,0,temp.length);
        }
        return ret;
    }
}

File: ios/ofdpdfsigner_ios/ofdpdfsignerlib/OFDStamp.swift
//
//  OFDStamp.swift
//  ofdcreatorlib
//
//  Created by zizheng fan on 2024/6/7.
//

import Foundation

//----------------------------------------------------------------------------------------
typealias Type_OFDSignStamp_Pos_Handle = @convention(c) (UnsafeMutableRawPointer,UnsafeRawPointer?, Int32,Int32,Double,Double,Double,Double,Bool, Type_CallBack_Sign,Type_CallBack_getUserCert,UnsafeRawPointer?, UnsafeMutablePointer<UnsafeMutablePointer<UInt8>?>?, UnsafeMutablePointer<size_t>?) -> Int32
private var Function_OFDSignStamp_Pos_Handle: Type_OFDSignStamp_Pos_Handle? = unsafeBitCast(dlsym(UnsafeMutableRawPointer(bitPattern: -2), "ofd_positionSign"), to: Type_OFDSignStamp_Pos_Handle.self)

typealias Type_OFDSignStamp_Riding_Handle = @convention(c) (UnsafeMutableRawPointer,UnsafeRawPointer?, Int32,Int32,Double,Double,Bool, Type_CallBack_Sign,Type_CallBack_getUserCert,UnsafeRawPointer?, UnsafeMutablePointer<UnsafeMutablePointer<UInt8>?>?, UnsafeMutablePointer<size_t>?) -> Int32
private var Function_OFDSignStamp_Riding_Handle: Type_OFDSignStamp_Riding_Handle? = unsafeBitCast(dlsym(UnsafeMutableRawPointer(bitPattern: -2), "ofd_ridingSign"), to: Type_OFDSignStamp_Riding_Handle.self)

typealias Type_OFDSignStamp_KeyWord_Handle = @convention(c) (UnsafeMutableRawPointer,UnsafeRawPointer?, Int32,UnsafePointer<CChar>?,UnsafePointer<CChar>?,Double,Double,Bool, Type_CallBack_Sign,Type_CallBack_getUserCert,UnsafeRawPointer?, UnsafeMutablePointer<UnsafeMutablePointer<UInt8>?>?, UnsafeMutablePointer<size_t>?) -> Int32
private var Function_OFDSignStamp_KeyWord_Handle: Type_OFDSignStamp_KeyWord_Handle? = unsafeBitCast(dlsym(UnsafeMutableRawPointer(bitPattern: -2), "ofd_keywordSign"), to: Type_OFDSignStamp_KeyWord_Handle.self)

typealias Type_OFDSignStamp_ClearSigns_Handle = @convention(c) (UnsafeMutableRawPointer,UnsafeMutablePointer<UnsafeMutablePointer<UInt8>?>?, UnsafeMutablePointer<size_t>?) -> Int32
private var Function_OFDSignStamp_ClearSigns_Handle: Type_OFDSignStamp_ClearSigns_Handle? = unsafeBitCast(dlsym(UnsafeMutableRawPointer(bitPattern: -2), "ofd_clearSigns"), to: Type_OFDSignStamp_ClearSigns_Handle.self)

//----------------------------------------------------------------------------------------
public class OFDStamp {
    
    public static let RIDING_LEFT : Int32 = 0;
    public static let RIDING_RIGHT : Int32 = 1;
    public static let RIDING_TOP : Int32 = 2;
    public static let RIDING_BOTTOM : Int32 = 3;
    public static let RIDING_CUTTING_RIGHT : Int32 = 4;
    public static let RIDING_CUTTING_TOP : Int32 = 5;
    
    /*
     OFD 位置盖章
     fileHandle:PDF文件句柄
     stampData：印章数据
     pageIndex：页码 0开始
     x：开始位置
     y：开始位置
     stampwidth：印章宽
     stampheight: 印章高
     isLocked：是否继续盖章
     signCallBack：签名回调
     返回：电子印章数据
     */
    public static func positionSign(fileHandle : UInt,sealData : [UInt8]?,pageNumber : Int32,stampX : Double,stampY : Double,
                                    width : Double, height : Double,isLocked : Bool,
                                    withClosure signCallback: @escaping Type_CallBack_Sign,withClosure getUserCertCallback: @escaping Type_CallBack_getUserCert) -> [UInt8]?
    {
        var retBytes : [UInt8]? = nil;
        if fileHandle == 0 {
            return retBytes;
        }
        if sealData == nil || sealData!.count == 0 {
            return retBytes;
        }
        if let handle = UnsafeMutableRawPointer(bitPattern: fileHandle) {
            if let function = Function_OFDSignStamp_Pos_Handle {
                //print("createSeal")
                let inDataPointer = sealData!.withUnsafeBytes({ $0 }).baseAddress
                let inDatalen = Int32(sealData!.count)
                
                var dataPtr: UnsafeMutablePointer<UInt8>? = nil
                var datalen: size_t = 0
                
                let result = function(handle,inDataPointer,inDatalen, pageNumber, stampX, stampY,width,height,!isLocked, signCallback,getUserCertCallback,nil,&dataPtr, &datalen)
                
                if result == 0 {
                    retBytes = Utils.convertUnsafePointerToArray(dataPtr: dataPtr, datalen: datalen)
                } else {
                    print("positionSign error:")
                    print(result)
                }
                
                if let pdataPtr = dataPtr {
                    Utils.freeHandle(handlePtr: pdataPtr)
                }
            }
        }

        return retBytes;
    }
    
    /*
     OFD 文档验证
     fileData：OFD数据
     verifyCallback：验证回调
     返回：验证结果，json数据
     */
    public static func verify(fileData : [UInt8]?, withClosure verifyCallback: @escaping Type_CallBack_Verify) -> String?
    {
        var ret : String = ""
        if fileData == nil || fileData!.count == 0 {
            return ret;
        }
        
        if let function = Function_VerifyDocment {
            //print("verifySeal")
            let inDataPointer = fileData!.withUnsafeBytes({ $0 }).baseAddress
            let inDatalen = Int32(fileData!.count)
            
            var dataPtr: UnsafeMutablePointer<UInt8>? = nil
            var datalen: size_t = 0
            
            let result = function(inDataPointer,inDatalen,1,"",verifyCallback,nil,&dataPtr, &datalen)
            
            if result == 0 {
                ret = String(data: Utils.convertToData(bytes: dataPtr, lenght: Int32(datalen))!, encoding: .utf8)!
            } else {
                print("verify error:")
                print(result)
            }
            
            if let pdataPtr = dataPtr {
                Utils.freeHandle(handlePtr: pdataPtr)
            }
            
        }
        
        return ret;
    }
    
    /*
     OFD文件骑缝章签章
     fileHandle:OFD文件句柄
     imageData：印章数据
     RidingType：骑缝章类型
     stampwidth：印章宽
     stampheight: 印章高
     isLocked：是否继续盖章
     signCallback：签名回调
     getUserCertCallback:用户证书
     返回：电子印章数据
     */

    public static func ridingSign(fileHandle : UInt,sealData : [UInt8]?,ridingType : String,
                                          width : Double, height : Double,isLocked : Bool,withClosure signCallback: @escaping Type_CallBack_Sign,withClosure getUserCertCallback: @escaping Type_CallBack_getUserCert)  -> [UInt8]?
    {
        var retBytes : [UInt8]? = nil;
        if fileHandle == 0 {
            return retBytes;
        }
        if sealData == nil || sealData!.count == 0 {
            return retBytes;
        }
        
        if let handle = UnsafeMutableRawPointer(bitPattern: fileHandle) {
            if let function = Function_OFDSignStamp_Riding_Handle {
                //print("createSeal")
                let inDataPointer = sealData!.withUnsafeBytes({ $0 }).baseAddress
                let inDatalen = Int32(sealData!.count)
                
                var dataPtr: UnsafeMutablePointer<UInt8>? = nil
                var datalen: size_t = 0
                
                let iType = Int32(ridingType)!
                
                let result = function(handle,inDataPointer,inDatalen, iType, width,height,!isLocked, signCallback,getUserCertCallback,nil,&dataPtr, &datalen)
                
                if result == 0 {
                    retBytes = Utils.convertUnsafePointerToArray(dataPtr: dataPtr, datalen: datalen)
                } else {
                    print("positionSign error:")
                    print(result)
                }
                
                if let pdataPtr = dataPtr {
                    Utils.freeHandle(handlePtr: pdataPtr)
                }
            }
        }

        return retBytes;
    }
    
  
    /*
     OFD文件骑缝章签章
     fileHandle:OFD文件句柄
     imageData：印章数据
     keyword：关键字
     KeywordRule：关键字规则
     stampwidth：印章宽
     stampheight: 印章高
     isLocked：是否继续盖章
     signCallback：签名回调
     getUserCertCallback:用户证书
     返回：电子印章数据
     */
    public static func keywordSign(fileHandle : UInt,sealData : [UInt8]?, keyWord : String,keyWordRules : String,
                                   width : Double, height : Double,isLocked : Bool,withClosure signCallback: @escaping Type_CallBack_Sign,withClosure getUserCertCallback: @escaping Type_CallBack_getUserCert) -> [UInt8]?
    {
        var retBytes : [UInt8]? = nil;
        if fileHandle == 0 {
            return retBytes;
        }
        if sealData == nil || sealData!.count == 0 {
            return retBytes;
        }
        
        if let handle = UnsafeMutableRawPointer(bitPattern: fileHandle) {
            if let function = Function_OFDSignStamp_KeyWord_Handle {
                //print("createSeal")
                let inDataPointer = sealData!.withUnsafeBytes({ $0 }).baseAddress
                let inDatalen = Int32(sealData!.count)
                
                var dataPtr: UnsafeMutablePointer<UInt8>? = nil
                var datalen: size_t = 0
                
              
                
                let result = function(handle,inDataPointer,inDatalen, keyWord,keyWordRules, width,height,!isLocked, signCallback,getUserCertCallback,nil,&dataPtr, &datalen)
                
                if result == 0 {
                    retBytes = Utils.convertUnsafePointerToArray(dataPtr: dataPtr, datalen: datalen)
                } else {
                    print("positionSign error:")
                    print(result)
                }
                
                if let pdataPtr = dataPtr {
                    Utils.freeHandle(handlePtr: pdataPtr)
                }
            }
        }

        return retBytes;
    }
    
    /*
     删除OFD文档中全部印章和数字签名
     fileHandle:OFD文件句柄
     返回：删除OFD文档中全部印章和数字签名后的文档数据
     */
    public static func clearSigns(fileHandle : UInt) -> [UInt8]?
    {
        var retBytes : [UInt8]? = nil;
        if fileHandle == 0 {
            return retBytes;
        }
        if let handle = UnsafeMutableRawPointer(bitPattern: fileHandle) {
            if let function = Function_OFDSignStamp_ClearSigns_Handle {
              
                var dataPtr: UnsafeMutablePointer<UInt8>? = nil
                var datalen: size_t = 0
                
                let result = function(handle,&dataPtr, &datalen)
                
                if result == 0 {
                    retBytes = Utils.convertUnsafePointerToArray(dataPtr: dataPtr, datalen: datalen)
                } else {
                    print("clearSigns error:")
                    print(result)
                }
                
                if let pdataPtr = dataPtr {
                    Utils.freeHandle(handlePtr: pdataPtr)
                }
            }
        }

        return retBytes;
    }
}
//----------------------------------------------------------------------------------------
